File: src/03_DuplicationInArray/03_DuplicationInArray.py
class Solution:
    def findRepeatNumber(self, nums: list[int], use_hash=False) -> int:
        if use_hash:
            return self.withHash(nums)
        else:
            return self.withSwap(nums)

    def withHash(self, nums: list[int]) -> int:
        dic = set() 
        for num in nums:
            if num in dic:
                return num
            dic.add(num)
        return -1

    def withSwap(self, nums: list[int]) -> int:
        i = 0
        while i < len(nums):
            if i == nums[i]:
                i += 1
                continue
            if nums[nums[i]] == nums[i]:
                return nums[i]
            # 注意赋值顺序
            nums[nums[i]], nums[i] = nums[i], nums[nums[i]]
        return -1


class Test:
    def __init__(self) -> None:
        self.solution = Solution()

    def checkResult(self, testName: str, nums: list[int], expected: list[int]) -> None:
        text = testName
        result = self.solution.findRepeatNumber(nums)
        if result in expected:
            text += " passed."
        else:
            text += " failed."
        print(text)

    # 重复的数字是数组中最小的数字
    def Test1(self):
        nums = [2, 1, 3, 1, 4]
        expected = [1]
        self.checkResult("Test1", nums, expected)

    # 重复的数字是数组中最大的数字
    def Test2(self):
        nums = [2, 4, 3, 1, 4]
        expected = [4]
        self.checkResult("Test2", nums, expected)

    # 数组中存在多个重复的数字
    def Test3(self):
        nums = [2, 4, 2, 1, 4]
        expected = [2, 4]
        self.checkResult("Test3", nums, expected)

    # 没有重复的数字
    def Test4(self):
        nums = [2, 1, 3, 0, 4]
        expected = [-1]  # not in use in the test function
        self.checkResult("Test4", nums, expected)

    # 没有重复的数字
    def Test5(self):
        nums = [2, 1, 3, 5, 4, 0]
        expected = [-1]  # not in use in the test function
        self.checkResult("Test5", nums, expected)

    # 无效的输入
    def Test6(self):
        nums = []
        expected = [-1]  # not in use in the test function
        self.checkResult("Test6", nums, expected)

    def runTest(self):
        self.Test1()
        self.Test2()
        self.Test3()
        self.Test4()
        self.Test5()
        self.Test6()


if __name__ == "__main__":
    test = Test()
    test.runTest()

File: src/03_DuplicationInArray/03_DuplicationInArray.cpp
// 面试题3：找出数组中重复的数字
// 题目：在一个长度为n的数组里的所有数字都在0到n-1的范围内。数组中某些数字是重复的，但不知道有几个数字重复了，
// 也不知道每个数字重复了几次。请找出数组中任意一个重复的数字。例如，如果输入长度为7的数组{2,
// 3, 1, 0, 2, 5, 3}， 那么对应的输出是重复的数字2或者3。

#include <iostream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>

using namespace std;

class Solution {
   public:
    int findRepeatNumber( vector<int>& nums, bool use_hash = false ) {
        if ( use_hash )
            return withHash( nums );
        else
            return withSwap( nums );
    }

   private:
    // 利用哈希表
    int withHash( vector<int>& nums ) {
        unordered_map<int, bool> dic;  // 用哈希表存储每个数是否存在
        for ( int num : nums ) {
            if ( dic[ num ] ) return num;  // 如果有重复数字，则提前返回
            dic[ num ] = true;
        }
        return -1;  // 如果无重复数字，则返回-1
    }

    // 原地交换。由于注意到数组长为n，而数字的取值范围都是[0,n-1]。
    // 说明数组元素的索引和值是一对多关系。
    // 通过遍历数组并通过交换操作，使元素与值一一对应。
    // 此时通过索引映射对应的值，起到与哈希表等价的作用
    int withSwap( vector<int>& nums ) {
        int i = 0;
        while ( i < nums.size() ) {
            // 此时数字已在索引位置，使索引后移，并跳过余下过程
            if ( nums[ i ] == i ) {
                i++;
                continue;
            }
            // 索引nums[i]和索引i的值相同，且两个索引不同，说明重复，返回
            if ( nums[ nums[ i ] ] == nums[ i ] ) return nums[ i ];
            // 其余情况交换索引nums[i]和i的值
            std::swap( nums[ i ], nums[ nums[ i ] ] );
        }
        return -1;
    }
};

class Test {
   public:
    void checkResult( const string testName, vector<int>& nums,
                      vector<int>& expected ) {
        cout << testName;
        int result = solution.findRepeatNumber( nums );
        if ( contain( expected, result ) )
            cout << " passed.\n";
        else
            cout << " failed.\n";
    }

    // 重复的数字是数组中最小的数字
    void Test1() {
        vector<int> nums{ 2, 1, 3, 1, 4 };
        vector<int> expected{ 1 };
        checkResult( "Test1", nums, expected );
    }

    // 重复的数字是数组中最大的数字
    void Test2() {
        vector<int> nums{ 2, 4, 3, 1, 4 };
        vector<int> expected{ 4 };
        checkResult( "Test2", nums, expected );
    }

    // 数组中存在多个重复的数字
    void Test3() {
        vector<int> nums{ 2, 4, 2, 1, 4 };
        vector<int> expected{ 2, 4 };
        checkResult( "Test3", nums, expected );
    }

    // 没有重复的数字
    void Test4() {
        vector<int> nums{ 2, 1, 3, 0, 4 };
        vector<int> expected{ -1 };  // not in use in the test function
        checkResult( "Test4", nums, expected );
    }

    // 没有重复的数字
    void Test5() {
        vector<int> nums{ 2, 1, 3, 5, 4, 0 };
        vector<int> expected{ -1 };  // not in use in the test function
        checkResult( "Test5", nums, expected );
    }

    // 无效的输入
    void Test6() {
        vector<int> nums;
        vector<int> expected{ -1 };  // not in use in the test function
        checkResult( "Test6", nums, expected );
    }

    void runTest() {
        Test1();
        Test2();
        Test3();
        Test4();
        Test5();
        Test6();
    }

   private:
    Solution solution;

    bool contain( vector<int>& nums, int num ) {
        for ( int n : nums ) {
            if ( num == n ) return true;
        }
        return false;
    }
};

int main() {
    Test test;
    test.runTest();
    return 0;
}